src/light.h

```diff
@@ -65,18 +65,17 @@ inline u8 decode_light(u8 light)
 // 0.0 <= return value <= 1.0
 inline float decode_light_f(float light_f)
 {
-	s32 i = (u32)(light_f * LIGHT_MAX + 0.5);
-
-	if (i <= 0)
-		return (float)light_decode_table[0] / 255.0;
-	if (i >= LIGHT_SUN)
-		return (float)light_decode_table[LIGHT_SUN] / 255.0;
-
-	float v1 = (float)light_decode_table[i - 1] / 255.0;
-	float v2 = (float)light_decode_table[i] / 255.0;
-	float f0 = (float)i - 0.5;
-	float f = light_f * LIGHT_MAX - f0;
-	return f * v2 + (1.0 - f) * v1;
+	light_f *= LIGHT_SUN; // should be [0, LIGHT_SUN]
+	if (light_f <= 0)
+		return 0.0f;
+	if (light_f >= LIGHT_SUN)
+		return 1.0f;
+	u8 index = light_f; // truncate fractional part
+	assert(index < LIGHT_SUN); // index <= light_f < LIGHT_SUN
+	float d = light_f - index;
+	float v1 = light_decode_table[index] / 255.0f;
+	float v2 = light_decode_table[index + 1] / 255.0f;
+	return d * v2 + (1.0f - d) * v1;
```

Comment:
One multiplication less: `d * (v2 - v1) + v1`

Reply:
But less clear IMO. Do you write linear interpolation as θ (b − a) + a?

Reply:
SmallJoker's suggestion still looks clear as an interpolation to me, i support that change. You could add a comment to make clear it is an interpolation between v1 and v2.

Reply:
Well, these are two different ideas on how to interpolate:
   * `d * v2 + (1.0f - d) * v1`
     Take `v2` for the maximal value of `d` and `v1` for the minimal value of `d`; decreasing the higher `d` gets.
   * `d * (v2 - v1) + v1`
     Take the delta of the minimum and maximal value and add it to the minimal value proportional to `d`

Logically both are equivalent, but I would prefer the shorter one which is also available as noise helper function:
https://github.com/minetest/minetest/blob/a4f41e7bfc6c2d823eab1465258617cb033a452c/src/noise.cpp#L196-L199